[blog] Add Material UI v6 stable release (#41932)
mui/material-ui: docs/pages/blog/material-ui-v6-is-out.md

When light and dark color schemes are enabled, Material UI v6 uses the CSS media `prefers-color-scheme` as the default method for switching between modes.

This method works with server-side rendering without flickering and is compatible with all modern browsers.

### A hook for toggling between modes

The new `useColorScheme` hook lets you read and update the mode in your application.

```jsx
import { useColorScheme } from '@mui/material/styles';

function ModeSwitcher() {
  const { mode, setMode } = useColorScheme();
  if (!mode) return null;
  return (
    <select onChange={(event) => setMode(event.target.mode)}>
      <option value="system">System</option>
      <option value="light">Light</option>
      <option value="dark">Dark</option>
    </select>
  );
}
```

## CSS theme variables

The `cssVariables` flag converts serializable theme values like palette, spacing, typography, etc., into CSS variables.

```js
createTheme({ cssVariables: true, ... });
```

{{"component": "components/blog/material-ui-v6-is-out/ThemeTokens.js"}}

With CSS variables, you can easily integrate your preferred styling solution with Material UI.

```css title="styles.css"
.custom-card {

>   background-color: var(--palette-background-default);

Reviewer: Is it a type that the "mui" is not present in these variables. It is a bit conofusing as you read it, as in the previous section all variables start with the mui prefix.

Author: Nice catch.

Since the integration is an opt-in feature, you can upgrade to v6 and continue using Emotion if you'd like. The link to the migration guide is at the bottom of the page.

<!-- would be nice if we have a link to a prototype page -->
<!-- a link to Next.js repo that render templates -->

### React Server Components

Once integrated with Pigment CSS, Material UI v6 provides a separate set of layout components, including `Grid`, `Container`, and `Stack`, that are compatible with React Server Components.

```jsx
import Box from '@mui/material-pigment-css/Box';
import Grid from '@mui/material-pigment-css/Grid';
import Container from '@mui/material-pigment-css/Container';
import Stack from '@mui/material-pigment-css/Stack';
```

### Built-in sx prop support

With Pigment CSS integration, all JSX elements support the `sx` prop out of the box.

```diff
- import Box from '@mui/material/Box';

- <Box component="img" sx={{ padding: 2 }} />
+ <img sx={{ padding: 2 }} />
```

:::success
Box component is **no** longer required to use the `sx` prop.
:::

## Built-in support for light and dark modes

Material UI v6 supports light and dark modes out of the box.

> The new `colorSchemes` node lets you create a light and dark color schemes for your application:

Reviewer 2: ```suggestion
The new `colorSchemes` node lets you create light and dark color schemes for your application:
```

```js
const theme = createTheme({ colorSchemes: { dark: true } }); // light is generated by default.

function App() {
  return <ThemeProvider theme={theme}>...</ThemeProvider>;
}
```

When color schemes are used, Material UI ships these features out of the box:

- CSS media `prefers-color-scheme` is used as the default method for switching between modes.

> - You can achieve instant transitions between modes with `disableTransitionOnChange` prop.

Reviewer 2: ```suggestion
- You can achieve instant transitions between modes with the `disableTransitionOnChange` prop.
```

> - Selected mode is stored in the local storage to persist the user's preference.

Reviewer 2: ```suggestion
- The selected mode is stored in local storage to persist the user's preference.
```

```js
createTheme({ cssVariables: true, ... });
```

{{"component": "components/blog/material-ui-v6-is-out/ThemeTokens.js"}}

With CSS variables, you can easily integrate your preferred styling solution with Material UI.

```css title="styles.css"
.custom-card {
  background-color: var(--mui-palette-background-default);
  color: var(--mui-palette-text-primary);
  padding: var(--mui-spacing-2);
  font: var(--mui-font-body1);
}
```

## New method of applying styles

The new API `theme.applyStyles` has been added for creating specific mode styles. It's designed to replace the `theme.palette.mode === 'dark'` condition and to work with Pigment CSS.

The snippet below will work for both Emotion and Pigment CSS.

```jsx
const StyledInput = styled(InputBase)(({ theme }) => ({
  padding: 10,
  width: '100%',
  borderBottom: `1px solid #eaecef`,
  ...theme.applyStyles('dark', {
    borderBottom: '1px solid #30363d',
  })
  '& input': {
    borderRadius: 4,
    backgroundColor: '#fff',
    ...theme.applyStyles('dark', {
      backgroundColor: '#0d1117',
    })
  },
}));
```

## Container Queries

> We added [container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries) utility based on the existing `theme.breakpoints` API.

Reviewer 2: ```suggestion
We've added a [container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries) utility based on the existing `theme.breakpoints` API.
```

This feature lets you define styles based on the parent container's width instead of the viewport.

```jsx
const Component = styled('div')(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(2),
  [theme.containerQueries.up('sm')]: {
    flexDirection: 'row',
  },
  [theme.containerQueries('sidebar').up('400px')]: {
    flexDirection: 'row',
  },
}));
```

It also works with the `sx` prop:

```jsx
<Card
  sx={{
    '@sm': {
      flexDirection: 'row',
    },
    '@400/sidebar': {
      flexDirection: 'row',
    },
  }}
/>
```

## Stabilized Grid v2

Grid v2 has been stabilized and is now using CSS gap for creating space between grid items.

> The responsive configuration has been consolidating into `size` and `offset` props for consistency.

Reviewer 2: ```suggestion
The responsive configuration has been consolidated into `size` and `offset` props for consistency.
```

```jsx
import Grid from '@mui/material/Grid2';

<Grid container>
  <Grid size={{ xs: 6, sm: 4, lg: 3 }} />
  <Grid size={{ xs: 6, sm: 4, lg: 3 }} />
  <Grid size={{ xs: 6, sm: 4, lg: 3 }} />
</Grid>;
```

## Revamping the free templates

Explore the new and enhanced [Material UI free templates](https://mui.com/material-ui/getting-started/templates/) to see these amazing features in action. We've fully revamped the templates to provide the perfect starting point for your project, whether you're adding sleek styles or using the template's sections.

{{"component": "components/blog/material-ui-v6-is-out/FreeTemplatesBento.js"}}

> ### Custom syles

Reviewer 2: ```suggestion
### Custom styles
```

The new custom styles we've added allow you to dive into the best approaches for customization or simply copy and paste some stylish elements.

{{"component": "components/blog/material-ui-v6-is-out/CustomThemeComparison.js"}}

Stay tuned, as many more updates are coming in the upcoming months! We'd love to hear your feedback on which templates or sections you'd like to see next.

> For your full-stack needs you can also explore [Toolpad Core](https://mui.com/toolpad/)—a Next.js starter for building dashboards. It offers components like dashboard layout, next-auth powered sign-in page and more.

Reviewer 2: @prakhargupta1 I would imagine us adding a small subsection within the "Revamping the free templates" section, perhaps titled "Full stack apps", with an image similar to the one used in the "Custom styles" subsection - a composite of all the components we have available

```suggestion
### Full Stack Apps
To get started building full-stack apps fast, we're building [Toolpad Core](https://mui.com/toolpad) - get  Material UI, Next.js, routing, layout components and authentication out of the box while you build dashboards or internal tools.
```

I would place this subsection before the "Stay tuned, as many more..." line

Reviewer 2: I think a small image should be fine, given we're at the end of the page. People might skim the text entirely if there's no visual attention catcher.

---
title: Material UI v6 is out 🎉
description: We are so excited to announce the stable release of Material UI v6. It comes with Pigment CSS integration and a lot of improvements.
date: 2024-08-09T00:00:00.000Z
authors:
  [
    'aarongarciah',
    'brijeshb42',
    'diegoandai',
    'mnajdova',
    'samuelsycamore',
    'siriwatknp',
    'zanivan',
  ]
tags: ['Material UI', 'Product']
manualCard: true
---

> Material UI v6 is now stable! We're excited to share all the updates.

Author: @aarongarciah @DiegoAndai Would you mind adding React 19 compat?

> Material UI v6 is now stable! We're excited to share all the updates.

Reviewer 3: I'll add it.

> Material UI v6 is now stable! We're excited to share all the updates.

Reviewer 3: I added a "React 19 support" section before "Full-stack apps".

> 

Author: Added Table of contents and grouped the contents into `New features`, `Improvements`, etc.

> ## Opt-in CSS extraction via Pigment CSS

Author: That's true, I'll move this section to last.

> ## Opt-in CSS extraction via Pigment CSS

Author: Addressed all of the suggestions

---
title: Material UI v6 is out 🎉
description: We are so excited to announce the stable release of Material UI v6. It comes with Pigment CSS integration and a lot of improvements.
date: 2024-08-09T00:00:00.000Z
authors:
  [
    'aarongarciah',
    'brijeshb42',
    'diegoandai',
    'mnajdova',
    'samuelsycamore',
    'siriwatknp',
    'zanivan',
  ]
tags: ['Material UI', 'Product']
manualCard: true
---

Material UI v6 is now stable! We're excited to share all the updates.

## Table of contents

- [New features](#new-features)
  - [CSS theme variables](#css-theme-variables)
  - [Color schemes](#color-schemes)
  - [CSS media `prefers-color-scheme`](#css-media-prefers-color-scheme)
  - [Container Queries](#container-queries)
  - [New method of applying styles](#new-method-of-applying-styles)
- [Improvements](#improvements)
  - [Revamping the free templates](#revamping-the-free-templates)
  - [Stabilized Grid v2](#stabilized-grid-v2)
- [Experimental CSS extraction via Pigment CSS](#experimental-css-extraction-via-pigmentcss)
  - [React Server Components](#react-server-components)
  - [Built-in sx prop support](#built-in-sx-prop-support)

> - [Package size reduction](#package-size-reduction)

Reviewer 3: Could "Package size reduction" be places under "Improvements"?

Author: 👍 Ah, yes. That's better.

- [Full-stack apps](#full-stack-apps)
- [What's next](#whats-next)

## New features

### CSS theme variables

CSS Variables has become an industry standard for the web.
Material UI v6 introduces a new `cssVariables` flag that generates CSS variables from the serializable theme values like palette, spacing, typography, etc..

```js
const theme = createTheme({ cssVariables: true, ... });
```

> {{"component": "components/blog/material-ui-v6-is-out/ThemeTokens.js"}}

Reviewer 3: I'm seeing a double scroll bar (macOS with always visible scroll bars)

<img width="1019" alt="Screenshot 2024-08-19 at 12 26 05" src="https://github.com/user-attachments/assets/6fec2822-009b-4391-9192-54a16fa71ee2">

Author: Fixed.

---
title: Material UI v6 is out 🎉
description: We are so excited to announce the stable release of Material UI v6. It comes with Pigment CSS integration and a lot of improvements.
date: 2024-08-09T00:00:00.000Z
authors:
  [
    'aarongarciah',
    'brijeshb42',
    'diegoandai',
    'mnajdova',
    'romgrk',
    'samuelsycamore',
    'siriwatknp',
    'zanivan',
  ]
tags: ['Material UI', 'Product']
manualCard: true
---

Material UI v6 is now stable! We're excited to share all the updates.

## Table of contents

- [New features](#new-features)
  - [CSS theme variables](#css-theme-variables)
  - [Color schemes](#color-schemes)
  - [CSS media `prefers-color-scheme`](#css-media-prefers-color-scheme)
  - [Container Queries](#container-queries)
  - [New method of applying styles](#new-method-of-applying-styles)
- [Improvements](#improvements)
  - [Optimized runtime performance](#optimized-runtime-performance)
  - [Revamping the free templates](#revamping-the-free-templates)
  - [Stabilized Grid v2](#stabilized-grid-v2)
  - [Package size reduction](#package-size-reduction)

> - [Experimental CSS extraction via Pigment CSS](#experimental-css-extraction-via-pigmentcss)

Reviewer 4: This link is missing a dash between `pigment` and `css`.
```suggestion
- [Experimental CSS extraction via Pigment CSS](#experimental-css-extraction-via-pigment-css)
```

  '& input': {
    borderRadius: 4,
    backgroundColor: theme.palette.mode === 'dark' ? '#0d1117' : '#fff',
  },
}));
```

</codeblock>

:::info
We provide a codemod to help you migrate from `theme.palette.mode === 'dark'` to `theme.applyStyles()`. You can find the link to the migration guide at the end of the blog post.
:::

## Improvements

### Optimized runtime performance

[Several optimizations](https://github.com/mui/material-ui/pulls?q=is%3Apr+author%3Aromgrk+is%3Aclosed+perf+sort%3Aupdated-desc) have been made to improve the runtime performance of Material UI v6.

This is just the beginning of our performance improvements. There are more optimizations in our pipeline, so stay tuned for v6.x updates and the benchmark result!

### Revamping the free templates

Explore the new and enhanced [Material UI free templates](https://mui.com/material-ui/getting-started/templates/) to see these amazing features in action. We've fully revamped the templates to provide the perfect starting point for your project, whether you're adding sleek styles or using the template's sections.

{{"component": "components/blog/material-ui-v6-is-out/FreeTemplatesBento.js"}}

#### Custom styles

> The new custom styles we've added allow you to dive into the best approaches for customization or simply copy and paste some stylish elements.

Author: The templates come with a custom theme by default to showcase the theming feature of Material UI, I think that's what @zanivan wants to highlight.

https://next.mui.com/material-ui/getting-started/templates/dashboard/

date: 2024-08-09T00:00:00.000Z
authors:
  [
    'aarongarciah',
    'brijeshb42',
    'diegoandai',
    'mnajdova',
    'romgrk',
    'samuelsycamore',
    'siriwatknp',
    'zanivan',
  ]
tags: ['Material UI', 'Product']
manualCard: true
---

Material UI v6 is now stable! We're excited to share all the updates.

## Table of contents

- [New features](#new-features)
  - [CSS theme variables](#css-theme-variables)
  - [Color schemes](#color-schemes)
  - [CSS media `prefers-color-scheme`](#css-media-prefers-color-scheme)
  - [Container Queries](#container-queries)
  - [New method of applying styles](#new-method-of-applying-styles)
- [Improvements](#improvements)
  - [Optimized runtime performance](#optimized-runtime-performance)
  - [Revamping the free templates](#revamping-the-free-templates)
  - [Stabilized Grid v2](#stabilized-grid-v2)
  - [Package size reduction](#package-size-reduction)
- [Experimental CSS extraction via Pigment CSS](#experimental-css-extraction-via-pigmentcss)
  - [React Server Components](#react-server-components)
  - [Built-in sx prop support](#built-in-sx-prop-support)
- [React 19 support](#react-19-support)

> - [Full-stack apps](#full-stack-apps)

Reviewer 2: ```suggestion
- [Full stack apps](#full-stack-apps)
```

Reviewer 5: We do advise "server-side" and 'client-side' with a hyphen though: https://github.com/mui/material-ui/blob/8ab2137281267ad4fa766e4bd64d03baa5d44820/docs/mui-vale/styles/MUI/CorrectReferenceAllCases.yml#L30-L31

Maybe we need to add some rules for front(-)end, back(-)end and full(-)stack?

Then, if you're ready to integrate Pigment CSS into your project, follow the migration guide at the bottom of the page
:::

### React Server Components

Once integrated with Pigment CSS, Material UI v6 provides a separate set of layout components, including `Grid`, `Container`, and `Stack`, that are compatible with React Server Components.

```jsx
import Grid from '@mui/material-pigment-css/Grid';
import Container from '@mui/material-pigment-css/Container';
import Stack from '@mui/material-pigment-css/Stack';
```

### Built-in sx prop support

With Pigment CSS integration, all JSX elements support the `sx` prop out of the box.
The `Box` component is no longer required to use the `sx` prop.

```diff
-import Box from '@mui/material/Box';

-<Box component="img" sx={{ padding: 2 }} />
+<img sx={{ padding: 2 }} />
```

## React 19 support

Material UI v6 is ready for React 19. We've been testing Material UI with the latest [React 19 RC](https://react.dev/blog/2024/04/25/react-19) versions to ensure compatibility once the stable release of React 19 is out.

> ## Full-stack apps

Reviewer 5: How do you feel about just "Toolpad Core" as a title? Would it spark more interest?

```suggestion
## Toolpad Core
```

> To get started building full-stack apps fast, we're building [Toolpad Core](https://mui.com/toolpad/) - get Material UI, Next.js, routing, layout components and authentication out of the box while you build dashboards or internal tools.

Reviewer 2: ```suggestion
To get started building full stack apps fast, we're building [Toolpad Core](https://mui.com/toolpad/) - get Material UI, Next.js, routing, layout components and authentication out of the box while you build dashboards or internal tools.
```

Then, if you're ready to integrate Pigment CSS into your project, follow the migration guide at the bottom of the page
:::

### React Server Components

Once integrated with Pigment CSS, Material UI v6 provides a separate set of layout components, including `Grid`, `Container`, and `Stack`, that are compatible with React Server Components.

```jsx
import Grid from '@mui/material-pigment-css/Grid';
import Container from '@mui/material-pigment-css/Container';
import Stack from '@mui/material-pigment-css/Stack';
```

### Built-in sx prop support

With Pigment CSS integration, all JSX elements support the `sx` prop out of the box.
The `Box` component is no longer required to use the `sx` prop.

```diff
-import Box from '@mui/material/Box';

-<Box component="img" sx={{ padding: 2 }} />
+<img sx={{ padding: 2 }} />
```

## React 19 support

Material UI v6 is ready for React 19. We've been testing Material UI with the latest [React 19 RC](https://react.dev/blog/2024/04/25/react-19) versions to ensure compatibility once the stable release of React 19 is out.

> Also, we're working on backporting React 19 support to Material UI v5. Stay tuned.

Reviewer 3: @samuelsycamore I added this paragraph, would you mind a quick review?

---
title: Material UI v6 is out now 🎉
description: Material UI v6 ships with support for CSS variables, container queries, and advanced color schemes, as well as improved runtime performance and a reduced bundle size.

> date: 2024-08-09T00:00:00.000Z

Reviewer 3: ```suggestion
date: 2024-08-22T00:00:00.000Z
```

Reviewer 6: Lets set it to Monday and target to release that day

```suggestion
date: 2024-08-26T00:00:00.000Z
```

authors:
  [
    'aarongarciah',
    'brijeshb42',
    'diegoandai',
    'mnajdova',
    'romgrk',
    'samuelsycamore',
    'siriwatknp',
    'zanivan',
  ]
tags: ['Material UI', 'Product']
manualCard: true
---

> Material UI v6 is now stable! It comes with new features, improvements, and an experimental integration for static CSS extraction.

Reviewer 6: Nice! I'll add it

See [Package Phobia](https://packagephobia.com/result?p=@mui/material) for more details.
Instead of UMD, we recommend using ESM-based CDNs such as [esm.sh](https://esm.sh/).
For alternative installation methods, refer to the [CDN documentation](/material-ui/getting-started/installation/#cdn).

## Experimental CSS extraction via Pigment CSS

Material UI v5 uses Emotion as its default styling solution.
Because it's a runtime CSS-in-JS library, Emotion forces us to make compromises to performance and bundle size.

Material UI v6 introduces opt-in integration with [Pigment CSS](https://github.com/mui/pigment-css), our new zero-runtime styling library that eliminates the runtime overhead while preserving the API design patterns you already know and love.

:::warning
Pigment CSS integration is still in the experimental stage.
We recommend testing it out with our examples ([Next.js](https://github.com/mui/material-ui/tree/master/examples/material-ui-pigment-css-nextjs-ts) or [Vite](https://github.com/mui/material-ui/tree/master/examples/material-ui-pigment-css-vite)), and we'd love to get your feedback to help us improve it.

> When you're ready to integrate Pigment CSS into your project, you can find complete details in [the guide to migrating to Pigment CSS](/material-ui/migration/migrating-to-pigment-css/).

Author: ```suggestion
When you're ready to integrate Pigment CSS into your project, you can find complete details in [the guide to migrating to Pigment CSS](/material-ui/migration/migrating-to-pigment-css/).
```
